packages/core/src/types.ts

```diff
@@ -1259,239 +1252,53 @@ export interface MachineImplementationsSimplified<
   delays: DelayFunctionMap<TContext, TEvent, TAction>;
 }
 
-type MaybeNarrowedEvent<TIndexedEvents, TCausingLookup, K> = Cast<
-  Prop<
-    TIndexedEvents,
-    K extends keyof TCausingLookup
-      ? TCausingLookup[K]
-      : TIndexedEvents[keyof TIndexedEvents]
-  >,
-  EventObject
->;
-
-type MachineImplementationsActions<
-  TContext extends MachineContext,
-  TResolvedTypesMeta,
-  TEventsCausingActions = Prop<
-    Prop<TResolvedTypesMeta, 'resolved'>,
-    'eventsCausingActions'
-  >,
-  TIndexedEvents = Prop<Prop<TResolvedTypesMeta, 'resolved'>, 'indexedEvents'>,
-  TIndexedActors = Prop<Prop<TResolvedTypesMeta, 'resolved'>, 'indexedActors'>,
-  TIndexedActions = Prop<
-    Prop<TResolvedTypesMeta, 'resolved'>,
-    'indexedActions'
-  >,
-  TIndexedGuards = Prop<Prop<TResolvedTypesMeta, 'resolved'>, 'indexedGuards'>,
-  TIndexedDelays = Prop<Prop<TResolvedTypesMeta, 'resolved'>, 'indexedDelays'>,
-  TEmitted = Prop<Prop<TResolvedTypesMeta, 'resolved'>, 'emitted'>
-> = {
-  [K in keyof TIndexedActions]?: ActionFunction<
-    TContext,
-    MaybeNarrowedEvent<TIndexedEvents, TEventsCausingActions, K>,
-    Cast<Prop<TIndexedEvents, keyof TIndexedEvents>, EventObject>,
-    GetParameterizedParams<Cast<TIndexedActions[K], ParameterizedObject>>,
-    Cast<Prop<TIndexedActors, keyof TIndexedActors>, ProvidedActor>,
-    Cast<Prop<TIndexedActions, keyof TIndexedActions>, ParameterizedObject>,
-    Cast<Prop<TIndexedGuards, keyof TIndexedGuards>, ParameterizedObject>,
-    Cast<
-      Prop<TIndexedDelays, keyof TIndexedDelays>,
-      ParameterizedObject
-    >['type'],
-    Cast<TEmitted, EventObject>
+type MachineImplementationsActions<TTypes extends StateMachineTypes> = {
+  [K in TTypes['actions']['type']]?: ActionFunction<
+    TTypes['context'],
+    TTypes['events'],
+    TTypes['events'],
+    (TTypes['actions'] & { type: K })['params'],
```

Comment:
we had problems with this pattern elsewhere, see: https://github.com/statelyai/xstate/pull/4568/files#diff-af7184a4b0cdcc69ce20eda13e1aceda04efde70d1b26482886cf31e245ad049R40-R46